src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs

```diff
@@ -78,8 +92,69 @@ public async Task SaveChangesAsync(IHasAssociations[] owners)
                 source.Associations.Patch(target.Associations, associationComparer, (sourceAssociation, targetAssociation) => sourceAssociation.Patch(targetAssociation));
 
                 await repository.UnitOfWork.CommitAsync();
+                //Reset cached catalogs and catalogs
+                CatalogCacheRegion.ExpireRegion();
+            }
+
+        }
+
+        public async Task UpdateAssociationSetAsync(string ownerId, ProductAssociation[] associations)
+        {
+            var changedEntities = new List<AssociationEntity>();
+
+            var dbAssociations = associations.Select(x => AbstractTypeFactory<AssociationEntity>.TryCreateInstance().FromModel(x)).ToArray();
+            foreach (var dbAssociation in dbAssociations)
+            {
+                dbAssociation.ItemId = ownerId;
+            }
+            changedEntities.AddRange(dbAssociations);
+
+
+            using (var repository = _repositoryFactory())
+            {
+                var existEntities = await repository.Associations.Where(x => ownerId == x.ItemId).ToArrayAsync();
+
+                var target = new { Associations = new ObservableCollection<AssociationEntity>(existEntities) };
+                var source = new { Associations = new ObservableCollection<AssociationEntity>(changedEntities) };
+
+                var associationComparer = AnonymousComparer.Create((AssociationEntity x) => x.ItemId + ":" + x.AssociationType + ":" + x.AssociatedItemId + ":" + x.AssociatedCategoryId);
+                source.Associations.Patch(target.Associations, associationComparer, (sourceAssociation, targetAssociation) => sourceAssociation.Patch(targetAssociation));
+
+                var forRemove = existEntities.Where(c => !target.Associations.Any(t => t.Id == c.Id));
+                await repository.RemoveAssociationsAsync(forRemove.Select(c => c.Id).ToArray());
+                await repository.UnitOfWork.CommitAsync();
+                //Reset cached catalogs and catalogs
+                CatalogCacheRegion.ExpireRegion();
+            }
+        }
+
+        public async Task UpdateAssociationAsync(string ownerId, ProductAssociation association)
```

Comment:
Does it makes the same as UpdateAssociationSetAsync?

Reply:
answered here https://github.com/VirtoCommerce/vc-module-catalog/pull/356#discussion_r423538149